Fix hashjoin read after reordering

ylw510 opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)



### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
This PR is intended to resolve this issue: #85976 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


<!-- ch-version-info:start -->
### Version info
- Merged into: `26.1.1.448`
<!-- ch-version-info:end -->

ylw510 committed:
fix hashjoin read after reordering

ylw510 committed:
Update 03604_fix_hashjoin_read_after_reordering.sh

ylw510 committed:
Update 03604_fix_hashjoin_read_after_reordering.reference

nickitat commented on src/Interpreters/HashJoin/HashJoin.cpp:
```diff
@@ -1696,18 +1696,42 @@ void HashJoin::tryRerangeRightTableDataImpl(Map & map [[maybe_unused]])
             auto & columns = columns_list.back().columns;
             size_t start_row = columns.at(0)->size();
-            for (; it.ok(); ++it)
+            size_t new_rows = 0;
```
Could you pls elaborate on what the issue was exactly, and how this change will solve it?

nickitat committed:
Merge branch 'master' into fix_issue_85976

nickitat committed:
fix test

alesapin committed:
Merge remote-tracking branch 'origin/master' into HEAD

nickitat committed:
Merge branch 'master' into fix_issue_85976

nickitat committed:
Merge branch 'master' into fix_issue_85976

nickitat committed:
Merge branch 'master' into fix_issue_85976

nickitat committed:
fix

ylw510 committed:
Add unit test for RowRefList insert on empty list

Cover the case where RowRefList is default-constructed (rows == 0)
and the first insert initializes the root element instead of creating
a Batch, preventing a potential crash.
